src/test/java/io/pravega/connectors/flink/FlinkPravegaWriterITCase.java

```diff
@@ -311,37 +223,66 @@ public void testExactlyOnceWriter() throws Exception {
     @Test
     public void testExactlyOnceWriterWithWatermark() throws Exception {
         int numElements = 2000;
-
-        // set up the stream
         final String streamName = RandomStringUtils.randomAlphabetic(20);
+
+        final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment()
+                .setParallelism(1)
+                .enableCheckpointing(1000, CheckpointingMode.EXACTLY_ONCE);
+        env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime);
+        env.getConfig().setAutoWatermarkInterval(100);
+        env.getCheckpointConfig().enableUnalignedCheckpoints();
+        env.setRestartStrategy(RestartStrategies.fixedDelayRestart(1, 0L));
+
+        final FlinkPravegaWriter<Integer> sink = FlinkPravegaWriter.<Integer>builder()
+                .forStream(streamName)
+                .withPravegaConfig(SETUP_UTILS.getPravegaConfig())
+                .withSerializationSchema(new IntSerializer())
+                .withWriterMode(PravegaWriterMode.EXACTLY_ONCE)
+                .withTxnLeaseRenewalPeriod(Time.seconds(30))
+                .enableWatermark(true)
+                .build();
+
+        env
+                .addSource(new ThrottledIntegerGeneratingSource(numElements).withWatermarks(20))
+                .map(new FailingMapper<>(numElements / 2))
+                .addSink(sink).setParallelism(2);
+
+        writeAndCheckData(numElements, streamName, env, false, 1000);
+        checkWatermark(streamName);
+    }
+
+    // ----------------------------------------------------------------------------
+
+    private static class IntSerializer implements SerializationSchema<Integer> {
+
+        @Override
+        public byte[] serialize(Integer integer) {
+            return ByteBuffer.allocate(4).putInt(0, integer).array();
+        }
+    }
+
+    /**
+     * Write from 1 to numElements and check the result concurrently.
+     *
+     * @param numElements        The max num written by the function.
+     * @param streamName         The test stream name containing the data to be verified.
+     * @param env                The Flink environment to be executed, with the data stream configured.
+     * @param allowDuplicate     Check data in AT_LEAST_ONCE or EXACTLY_ONCE mode.
+     * @param waitSeconds        The maximum seconds we wait for the checker.
+     * @throws Exception on any errors.
+     */
+    void writeAndCheckData(int numElements,
```

Comment:
It's better if we can align all the tests with the same numElements and waitSeconds